compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs

```diff
@@ -87,61 +85,33 @@ createProjectPackageDb opts thisSdkVer deps0 dataDeps = do
     -- deal with data imports first
     let (fpDars, fpDalfs) = partition ((== ".dar") . takeExtension) dataDeps
     dars <- mapM extractDar fpDars
-    let dalfs = concatMap edDalfs dars
-    -- when we compile packages with different sdk versions or with dalf dependencies, we
-    -- need to generate the interface files
+    -- These are the dalfs that are in a DAR that has been passed in via data-dependencies.
     let dalfsFromDars =
             [ ( dropExtension $ takeFileName $ ZipArchive.eRelativePath e
-              , BSL.toStrict $ ZipArchive.fromEntry e)
-            | e <- dalfs
+              , BSL.toStrict $ ZipArchive.fromEntry e
+              )
+            | e <- concatMap edDalfs dars
             ]
+    -- These are dalfs that have been passed in directly as DALFs via data-dependencies.
     dalfsFromFps <-
         forM fpDalfs $ \fp -> do
-            bs <- B.readFile fp
+            bs <- BS.readFile fp
             pure (dropExtension $ takeFileName fp, bs)
     let allDalfs = dalfsFromDars ++ dalfsFromFps
-    pkgs <-
-        forM allDalfs $ \(name, dalf) -> do
-            (pkgId, package) <-
-                either (fail . DA.Pretty.renderPretty) pure $
-                Archive.decodeArchive Archive.DecodeAsMain dalf
-            pure
-                ( pkgId
-                , package
-                , dalf
-                , stringToUnitId $ parseUnitId name pkgId)
-    -- mapping from package id's to unit id's. if the same package is imported with
-    -- different unit id's, we would loose a unit id here.
-    let pkgMap =
-            MS.fromList
-                [(pkgId, unitId) | (pkgId, _pkg, _bs, unitId) <- pkgs]
-    -- order the packages in topological order
-    let (depGraph, vertexToNode, _keyToVertex) =
-            graphFromEdges $ do
-                (pkgId, dalf, bs, unitId) <- pkgs
-                let pkgRefs =
-                        [ pid
-                        | LF.PRImport pid <- toListOf packageRefs dalf
-                        ]
-                let getUid = getUnitId unitId pkgMap
-                let src = generateSrcPkgFromLf getUid (Just "Sdk") dalf
-                let templInstSrc =
-                        generateTemplateInstancesPkgFromLf
-                            getUid
-                            (Just "Sdk")
-                            pkgId
-                            dalf
-                pure
-                    ( (src, templInstSrc, unitId, dalf, bs)
-                    , pkgId
-                    , pkgRefs)
-    let pkgIdsTopoSorted = reverse $ topSort depGraph
+    pkgs <- forM allDalfs $ \(name, dalf) -> do
+        (pkgId, package) <-
+            either (fail . DA.Pretty.renderPretty) pure $
```

Comment:
If possible extract `either (fail . DA.Pretty.renderPretty) pure` as a separate function (like `requiredE` from the daml-assistant codebase).

```suggestion
            requiredE $
```

Reply:
I’ll leave it for now but I agree that we should factor this out eventually.